types/ldapjs-client/index.d.ts

```diff
@@ -0,0 +1,92 @@
+// Type definitions for ldapjs-client 0.1
+// Project: https://github.com/zont/ldapjs-client#readme
+// Definitions by: Valerio Coltrè <https://github.com/colthreepv>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// TypeScript Version: 2.2
+
+export class Filter {
+    matches(obj: any): boolean;
+    type: string;
+}
+
+export interface Change {
+    operation: string;
+    modification: {
+        [key: string]: any;
+    };
+}
+
+export interface SearchOptions {
+    scope: string;
+    filter: string | Filter;
+    attributes: string[];
+    sizeLimit: number;
+    timeLimit: number;
+    typesOnly: boolean;
+}
+export type PartialSearchOptions = Partial<SearchOptions>;
+
+export interface ClientOptions {
+    url: string;
+    tlsOptions?: object;
+    timeout?: number;
+}
+
+export class LdapClient {
+    constructor(options: ClientOptions);
+
+    /**
+     * Adds an entry to the LDAP server.
+     */
+    add(dn: string, entry: object): Promise<any>;
+
+    /**
+     * Performs a simple authentication against the server.
+     */
+    bind(dn: string, password: string): Promise<any>;
+
+    /**
+     * Deletes an entry from the LDAP server.
+     */
+    del(dn: string): Promise<any>;
+
+    /**
+     * Disconnect from the LDAP server and do not allow reconnection.
+     *
+     * If the client is instantiated with proper reconnection options, it's
+     * possible to initiate new requests after a call to unbind since the client
+     * will attempt to reconnect in order to fulfill the request.
+     *
+     * Calling destroy will prevent any further reconnection from occurring.
+     */
+    destroy(): Promise<void>;
+
+    /**
+     * Performs an LDAP modify against the server.
+     *
+     * @param dn the DN of the entry to modify.
+     * @param change update to perform (can be [Change]).
+     */
+    modify(dn: string, change: Change): Promise<any>;
+
+    /**
+     * Performs an LDAP modifyDN against the server.
+     */
+    modifyDN(dn: string, newName: string): Promise<any>;
+
+    /**
+     * Performs an LDAP search against the server.
+     *
+     * Note that the defaults for options are a 'base' search.
+     */
+    // tslint:disable-next-line: no-unnecessary-generics
+    search<T extends any>(base: string, options: PartialSearchOptions): Promise<T[]>;
```

Comment:
I think this is legitimate - the generic is really acting as a type assertion. You should instead pick `unknown[]` and let users cast.